Wrote additional high level API assembly functions

amartinhuertas opened the pull request.

fverdugo commented on src/FESpaces/Assemblers.jl:
```diff
@@ -270,44 +270,88 @@ function assemble_matrix(f::Function,a::Assembler,U::FESpace,V::FESpace)
   assemble_matrix(a,collect_cell_matrix(U,V,f(u,v)))
 end
 
+function assemble_matrix!(A,f::Function,a::Assembler,U::FESpace,V::FESpace)
```
Start the argument list with ::Function objects so that we can use do-block syntax. Also type-annotate the new array arguments, I.e.
```julia
assemble_matrix!(f::Function,A::AbstractMatrix,a::Assembler,U::FESpace,V::FESpace)
assemble_matrix_and_vector!(f::Function,b::Function,M::AbstractMatrix,r::AbstractVector,a::Assembler,U::FESpace,V::FESpace)
```
Apply this minor changes to all new introduced assembly functions.

amartinhuertas replied:
> Start the argument list with ::Function objects so that we can use do-block syntax

I though that the parameters that are modified in a `!`-like function should appear first in the list of parameters by convention. From your comment, I guess this is not strictly necessary, right?

fverdugo replied:
right, see e.g. `broadcast`

fverdugo reviewed (changes requested):
Hi @amartinhuertas thanks for the PR!

See my minor comments below.

fverdugo commented on src/FESpaces/Assemblers.jl:
```diff
@@ -270,44 +270,88 @@ function assemble_matrix(f::Function,a::Assembler,U::FESpace,V::FESpace)
   assemble_matrix(a,collect_cell_matrix(U,V,f))
 end
 
+function assemble_matrix!(A::AbstractMatrix,f,a::Assembler,U::FESpace,V::FESpace)
```
Function argument first here

amartinhuertas replied:
`f` is not a function, isnt?

fverdugo commented on src/FESpaces/Assemblers.jl:
```diff
@@ -270,44 +270,88 @@ function assemble_matrix(f::Function,a::Assembler,U::FESpace,V::FESpace)
   assemble_vector(a,collect_cell_vector(V,f))
 end
 
+function assemble_vector!(b::AbstractVector,f,a::Assembler,V::FESpace)
```
and here

amartinhuertas replied:
f is not a function, isnt?

fverdugo replied:
f is a callable object, usually a `Function`. 

to be consistent we have 2 main options:

- add  missing ::Function everywhere in this file, where we accept a function
- remove all ::Function anotations to allow generic callable objects

2 is more general, but 1 is less disruptive wrt what we have now and also works since the weak form is always defined via functions.

amartinhuertas replied:
I still dont understand why `f` it is function/callable object in this particular function. See, e.g., 

```
function collect_cell_vector(test::FESpace,a::DomainContribution)
  w = []
  r = []
  for trian in get_domains(a)
    cell_vec = get_contribution(a,trian)
    @assert ndims(eltype(cell_vec)) == 1
    cell_vec_r = attach_constraints_rows(test,cell_vec,trian)
    rows = get_cell_dof_ids(test,trian)
    push!(w,compress_contributions(cell_vec_r,trian))
    push!(r,compress_ids(rows,trian))
  end
  (w,r)
end
```

`a` is a callable object?

fverdugo replied:
You are right. It is not a function. it is an object for which `collect_cell_vector` and related functions are defined.

amartinhuertas replied:
ok, in any case I had to put the `f` and `b` args first in all function signatures to let all tests pass ...

fverdugo commented on src/FESpaces/Assemblers.jl:
```diff
@@ -270,44 +270,88 @@ function assemble_matrix(f::Function,a::Assembler,U::FESpace,V::FESpace)
   assemble_matrix_and_vector(a,collect_cell_matrix_and_vector(U,V,f,b))
 end
 
+function assemble_matrix_and_vector!(M,r,f,b,a::Assembler,U::FESpace,V::FESpace)
```
and here (and type-annotate array inputs)

amartinhuertas replied:
oops ... I forgot this one

amartinhuertas replied:
f is not a function, isnt?

fverdugo commented on src/FESpaces/Assemblers.jl:
```diff
@@ -270,44 +270,88 @@ function assemble_matrix(f::Function,a::Assembler,U::FESpace,V::FESpace)
   assemble_matrix(f,a,U,V)
 end
 
+function assemble_matrix!(A,f,U::FESpace,V::FESpace)
```
and here

fverdugo commented on src/FESpaces/Assemblers.jl:
```diff
@@ -270,44 +270,88 @@ function assemble_matrix(f::Function,a::Assembler,U::FESpace,V::FESpace)
   assemble_vector(f,a,V)
 end
 
+function assemble_vector!(b,f,V::FESpace)
```
and here

fverdugo commented on src/FESpaces/Assemblers.jl:
```diff
@@ -270,44 +270,88 @@ function assemble_matrix(f::Function,a::Assembler,U::FESpace,V::FESpace)
   assemble_matrix_and_vector(f,b,a,U,V)
 end
 
+function assemble_matrix_and_vector!(M,r,f,b,U::FESpace,V::FESpace)
```
and here

amartinhuertas commented:
Hi @fverdugo ... I already addressed all your comments. Anyway there is still something I do not understand. Some times we pass arguments to the `f` and `b` parameters which are not actually functions/callable objects ... I dont know perphaps I am confused. In any case, I think the PR is ready to be merged (if the tests pass).

fverdugo commented:
You are right. It is not a function. it is an object for which collect_cell_vector and related functions are defined. E.g. a DomainContribution
